[css-anchor-position-1] Implement parsing `@position-try` rule

tuankiet65 opened the pull request:
#### cc994137838b32d393c63fb549185375a96e2354
<pre>
[css-anchor-position-1] Implement parsing `@position-try` rule
<a href="https://rdar.apple.com/129789726">rdar://129789726</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=275455">https://bugs.webkit.org/show_bug.cgi?id=275455</a>

Reviewed by Antti Koivisto.

Implement parsing `@position-try` rule in CSS Anchor Positioning spec.
Note that the rules are only parsed and not yet collected for use in
`position-try-fallbacks`. This will be implemented in a future patch,
at least after parsing support for `position-try-fallbacks` is
completed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-allowed-declarations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-invalidation-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-parse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
  Added exception for CSSPositionTryRule, which inherits StyleRuleBase.
  StyleRuleBase doesn&apos;t define a virtual destructor, which trips up the
  checker. This is okay since StyleRuleBase defines its own delete operator
  that calls the correct destructor depends on the derived type. See
  <a href="https://github.com/WebKit/WebKit/pull/37285.">https://github.com/WebKit/WebKit/pull/37285.</a>
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
  Added exception for JSCSSPositionTryRule.cpp. Auto-generated file so can&apos;t
  be fixed except by fixing the code generator.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSPositionTryRule.cpp: Added.
(WebCore::StyleRulePositionTry::create):
(WebCore::StyleRulePositionTry::StyleRulePositionTry):
(WebCore::StyleRulePositionTry::protectedMutableProperties):
(WebCore::CSSPositionTryRule::create):
(WebCore::CSSPositionTryRule::CSSPositionTryRule):
(WebCore::CSSPositionTryRule::cssText const):
(WebCore::CSSPositionTryRule::reattach):
(WebCore::CSSPositionTryRule::name const):
(WebCore::CSSPositionTryRule::style):
* Source/WebCore/css/CSSPositionTryRule.h: Added.
(isType):
* Source/WebCore/css/CSSPositionTryRule.idl: Added.
* Source/WebCore/css/CSSProperty.cpp:
(WebCore::CSSProperty::isInsetProperty): Added some inset properties that we missed.
(WebCore::CSSProperty::isSizingProperty): Moved from AnchorPositionEvaluator.cpp
(WebCore::CSSProperty::isMarginProperty): Moved from AnchorPositionEvaluator.cpp
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::visitDerived):
(WebCore::StyleRuleBase::createCSSOMWrapper const):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isPositionTryRule const):
* Source/WebCore/css/StyleRuleType.h: Added new StyleRule type for @position-try.
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::traverseSubresources const):
(WebCore::StyleSheetContents::mayDependOnBaseURL const):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSAtRuleID.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeAtRule): Added case to consume @position-try rule.
(WebCore::CSSParserImpl::consumePositionTryRule): Added code to parse @position-try.
(WebCore::ruleDoesNotAllowImportant):
(WebCore::CSSParserImpl::consumeDeclaration): Returned early if current style rule does not
allow !important. Rejected custom properties when parsing @position-try.
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue): Added code to parse descriptors in @position-try rule.
(WebCore::propertyAllowedInPositionTryRule):
(WebCore::CSSPropertyParser::parsePositionTryDescriptor):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::evaluateSize):
(WebCore::Style::isSizingProperty): Moved to CSSProperty.cpp
(WebCore::Style::isMarginProperty): Moved to CSSProperty.cpp
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):

Canonical link: <a href="https://commits.webkit.org/287528@main">https://commits.webkit.org/287528@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1539c506b49394cd2e0d4cc93d114c5977b0f6ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79829 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58830 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33228 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84354 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30828 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81938 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67904 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7120 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62398 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20240 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82897 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52455 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72701 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42709 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49795 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26850 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29279 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27323 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85780 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7060 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4949 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70661 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7234 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68541 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69906 "70 api tests failed or timed out") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17448 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13908 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12823 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7019 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12592 "Found 3 new failures in JSWebCodecsEncodedVideoChunk.cpp, css/CSSPositionTryRule.h, JSCSSPositionTryRule.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6877 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10389 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8683 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSPositionTryRule.cpp:
```diff
@@ -0,0 +1,111 @@
+    StringBuilder builder;
+    builder.append("@position-try "_s);
+    builder.append(name());
+    builder.append(" {"_s);
```
```suggestion
    builder.append("@position-try "_s, name(), " {"_s);
```

It's more efficient to do coalesce calls as much as possible.

tuankiet65 replied:
I didn't know append can take an arbitrary number of arguments!

nt1m commented on Source/WebCore/css/CSSPositionTryRule.cpp:
```diff
@@ -0,0 +1,111 @@
+    if (auto declarations = m_positionTryRule->properties().asText(); !declarations.isEmpty()) {
+        builder.append(" "_s);
+        builder.append(declarations);
+        builder.append(" "_s);
```
```suggestion
        builder.append(' ', declarations, ' ');
```

nt1m commented on Source/WebCore/css/CSSPositionTryRule.cpp:
```diff
@@ -0,0 +1,111 @@
+        builder.append(declarations);
+        builder.append(" "_s);
+    } else
+        builder.append(" "_s);
```
```suggestion
        builder.append(' ');
```

nt1m commented on Source/WebCore/css/CSSPositionTryRule.cpp:
```diff
@@ -0,0 +1,111 @@
+    } else
+        builder.append(" "_s);
+
+    builder.append("}"_s);
```
```suggestion
    builder.append('}');
```

nt1m commented on Source/WebCore/css/CSSPositionTryRule.h:
```diff
@@ -0,0 +1,100 @@
+
+//     std::optional<ViewTransitionNavigation> m_navigation;
+//     Vector<AtomString> m_types;
+// };
```
nit: remove this.

tuankiet65 replied:
Yeah I had StyleRuleViewTransition there as a template to follow but forgot to remove it.

nt1m commented on Source/WebCore/css/CSSPositionTryRule.idl:
```diff
@@ -0,0 +1,37 @@
+interface CSSPositionTryRule : CSSRule {
+    readonly attribute [AtomString] CSSOMString name;
+    
+    // FIXME (FIXME: put a Bugzilla link): this should be a CSSPositionTryDescriptors.
```
Can you add the bugzilla link before landing?

nt1m commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -322,7 +323,16 @@ void CSSParserImpl::parseStyleSheetForInspector(const String& string, const CSSP
 {
-    if (!rule || allowedRules == CSSParserImpl::AllowedRules::FontFeatureValuesRules || allowedRules == CSSParserImpl::AllowedRules::KeyframeRules || allowedRules == CSSParserImpl::AllowedRules::CounterStyleRules || allowedRules == CSSParserImpl::AllowedRules::ViewTransitionRules || allowedRules == CSSParserImpl::AllowedRules::NoRules)
+    // FIXME: name suggestions?
+    bool ruleIsUnrestricted = !rule
```
ruleUsesAllowedRules ?

anttijk replied:
A general principle is to name booleans starting with a yes/no verb, like `isUnrestrictedRule`.

A switch might be better here.

nt1m replied:
Unrestricted is probably not the right verb fwiw, `allowedRules` are a set of restrictions too. Maybe just drop the variable and use a switch as Antti says.

tuankiet65 replied:
We dropped some `AllowRules` enums so the `if` condition is shorter now, hence I'll just keep the `if`. But yeah, I might make it into a switch if I have time.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -592,6 +595,33 @@ bool CSSPropertyParser::parseViewTransitionDescriptor(CSSPropertyID property)
+        || property == CSSPropertyJustifySelf
+        || property == CSSPropertyPlaceSelf
+        || property == CSSPropertyPositionAnchor;
+    // FIXME (needs bugzilla link): allow position-area when it's implemented
```
Just use the bugzilla link for position-area?

nt1m commented on Source/WebCore/style/RuleSetBuilder.cpp:
```diff
@@ -516,6 +519,10 @@ void RuleSetBuilder::addMutatingRulesToResolver()
+        if (auto* styleRulePositionTry = dynamicDowncast<StyleRulePositionTry>(rule.get())) {
+            // FIXME: do something here...
+            // FIXME: newline to make style checker happy for the time being
+        }
```
I would just remove this for now, and re-add it when you have an implementation of the evaluation ready.

nt1m replied:
I'd remove this for now.

tuankiet65 replied:
I removed it.

nt1m commented on Source/WebCore/css/parser/CSSParserImpl.h:
```diff
@@ -91,6 +92,7 @@ class CSSParserImpl {
         FontFeatureValuesRules,
         ViewTransitionRules,
+        PositionTryRules, // FIXME: is this needed??
         NoRules, // For parsing at-rules inside declaration lists (without nesting support)
```
CounterStyle/ViewTransition/PositionTry should actually just use `NoRules` since you can't nest further rules inside them. It is not the case however for the other ones.

tuankiet65 replied:
I decided to do some cleanup: https://github.com/WebKit/WebKit/pull/37163

anttijk commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -1043,6 +1054,31 @@ RefPtr<StyleRuleViewTransition> CSSParserImpl::consumeViewTransitionRule(CSSPars
+        return nullptr;
+
+    // FIXME: what's this?
+    if (m_observerWrapper) {
```
Something used for interacting with Inspector though details are hazy to me. (please don't land this FIXME)

tuankiet65 replied:
👍

nt1m commented on Source/WebCore/css/CSSPositionTryRule.h:
```diff
@@ -0,0 +1,81 @@
+public:
+    static Ref<StyleRulePositionTry> create(AtomString&& name, Ref<StyleProperties>&&);
+
+    // FIXME: more elegant way to express this?
```
You can probably drop this FIXME?

nt1m commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -1043,6 +1054,31 @@ RefPtr<StyleRuleViewTransition> CSSParserImpl::consumeViewTransitionRule(CSSPars
+    if (!m_context.propertySettings.cssAnchorPositioningEnabled)
+        return nullptr;
+
+    // prelude should ONLY be a <dashed-ident>
```
```suggestion
    // Prelude should only be a <dashed-ident>.
```

nt1m commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -1553,11 +1589,14 @@ bool CSSParserImpl::consumeDeclaration(CSSParserTokenRange range, StyleRuleType
         consumeCustomPropertyValue(range, variableName, important);
     }
 
+    // FIXME: might have to add StyleRuleType::PositionTry here?
```
You can't use !important in position-try from what I can understand, so probably makes sense to me to add it in there.

tuankiet65 replied:
I added it there.

nt1m commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -119,6 +119,8 @@ static RuleFlatteningStrategy flatteningStrategyForStyleRuleType(StyleRuleType s
     case StyleRuleType::FontFeatureValuesBlock:
     case StyleRuleType::FontPaletteValues:
+    // FIXME: probably should implement this?
+    case StyleRuleType::PositionTry:
```
Can you check if this works and remove the FIXME as appropriate? If it doesn't then please create a bug for it. If you don't know, I'd rather not include it.

tuankiet65 replied:
https://bugs.webkit.org/show_bug.cgi?id=284176

mdubet commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -56,6 +56,8 @@ class StyleRuleBase : public RefCounted<StyleRuleBase> {
 public:
     StyleRuleType type() const { return static_cast<StyleRuleType>(m_type); }
+
+    virtual ~StyleRuleBase() = default;
```
We should keep `StyleRuleBase` non virtual, and implement the trait system for it (cf the functions at the end of StyleRule.h) 

```
SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::StyleRulePositionTry)
    static bool isType(const WebCore::StyleRuleBase& rule) { return rule.isPositionTry(); }
SPECIALIZE_TYPE_TRAITS_END()
```

tuankiet65 replied:
Right, this is before I know `StyleRuleBase` is intentionally designed to be non-virtual

tuankiet65 replied:
I changed it back to the original.

tuankiet65 committed:
[css-anchor-position-1] Implement parsing `@position-try` rule
rdar://129789726
https://bugs.webkit.org/show_bug.cgi?id=275455

Reviewed by Antti Koivisto.

Implement parsing `@position-try` rule in CSS Anchor Positioning spec.
Note that the rules are only parsed and not yet collected for use in
`position-try-fallbacks`. This will be implemented in a future patch,
at least after parsing support for `position-try-fallbacks` is
completed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-allowed-declarations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-invalidation-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-parse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
  Added exception for CSSPositionTryRule, which inherits StyleRuleBase.
  StyleRuleBase doesn't define a virtual destructor, which trips up the
  checker. This is okay since StyleRuleBase defines its own delete operator
  that calls the correct destructor depends on the derived type. See
  https://github.com/WebKit/WebKit/pull/37285.
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
  Added exception for JSCSSPositionTryRule.cpp. Auto-generated file so can't
  be fixed except by fixing the code generator.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSPositionTryRule.cpp: Added.
(WebCore::StyleRulePositionTry::create):
(WebCore::StyleRulePositionTry::StyleRulePositionTry):
(WebCore::StyleRulePositionTry::protectedMutableProperties):
(WebCore::CSSPositionTryRule::create):
(WebCore::CSSPositionTryRule::CSSPositionTryRule):
(WebCore::CSSPositionTryRule::cssText const):
(WebCore::CSSPositionTryRule::reattach):
(WebCore::CSSPositionTryRule::name const):
(WebCore::CSSPositionTryRule::style):
* Source/WebCore/css/CSSPositionTryRule.h: Added.
(isType):
* Source/WebCore/css/CSSPositionTryRule.idl: Added.
* Source/WebCore/css/CSSProperty.cpp:
(WebCore::CSSProperty::isInsetProperty): Added some inset properties that we missed.
(WebCore::CSSProperty::isSizingProperty): Moved from AnchorPositionEvaluator.cpp
(WebCore::CSSProperty::isMarginProperty): Moved from AnchorPositionEvaluator.cpp
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::visitDerived):
(WebCore::StyleRuleBase::createCSSOMWrapper const):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isPositionTryRule const):
* Source/WebCore/css/StyleRuleType.h: Added new StyleRule type for @position-try.
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::traverseSubresources const):
(WebCore::StyleSheetContents::mayDependOnBaseURL const):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSAtRuleID.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeAtRule): Added case to consume @position-try rule.
(WebCore::CSSParserImpl::consumePositionTryRule): Added code to parse @position-try.
(WebCore::ruleDoesNotAllowImportant):
(WebCore::CSSParserImpl::consumeDeclaration): Returned early if current style rule does not
allow !important. Rejected custom properties when parsing @position-try.
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue): Added code to parse descriptors in @position-try rule.
(WebCore::propertyAllowedInPositionTryRule):
(WebCore::CSSPropertyParser::parsePositionTryDescriptor):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::evaluateSize):
(WebCore::Style::isSizingProperty): Moved to CSSProperty.cpp
(WebCore::Style::isMarginProperty): Moved to CSSProperty.cpp
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):

Canonical link: https://commits.webkit.org/287528@main
